File: examples/e2ee_chat.mjs
#!/usr/bin/env node
/**
 * Venice AI E2EE Chat Example
 *
 * Complete example of end-to-end encrypted communication with Venice AI API.
 * Your messages are encrypted client-side before transmission, and responses
 * are encrypted by the TEE before being sent back—Venice never sees plaintext.
 *
 * Requirements:
 *     npm install elliptic @noble/ciphers @noble/hashes
 *
 * Usage:
 *     export VENICE_API_KEY="your-api-key"
 *     node e2ee_chat.mjs
 */

import elliptic from 'elliptic';
import { gcm } from '@noble/ciphers/aes';
import { randomBytes } from '@noble/ciphers/webcrypto';
import { hkdf } from '@noble/hashes/hkdf';
import { sha256 } from '@noble/hashes/sha2';
import crypto from 'crypto';
import readline from 'readline';

const EC = elliptic.ec;

// Configuration
const API_KEY = process.env.VENICE_API_KEY;
const BASE_URL = process.env.VENICE_API_BASE_URL || 'https://api.venice.ai/api/v1';
const MODEL = 'e2ee-qwen3-30b-a3b'; // E2EE-capable model
const HKDF_INFO = 'ecdsa_encryption';

if (!API_KEY) {
  console.error('Error: VENICE_API_KEY environment variable not set');
  console.error('Usage: export VENICE_API_KEY="your-api-key"');
  process.exit(1);
}

/**
 * Convert bytes to hex string.
 */
function bytesToHex(bytes) {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

/**
 * Convert hex string to bytes.
 */
function hexToBytes(hex) {
  const h = hex.startsWith('0x') ? hex.slice(2) : hex;
  const bytes = new Uint8Array(h.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(h.substring(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/**
 * Generate ephemeral secp256k1 key pair for E2EE session.
 */
function generateKeyPair() {
  const ec = new EC('secp256k1');
  const keyPair = ec.genKeyPair();
  const privateKey = new Uint8Array(keyPair.getPrivate().toArray('be', 32));
  const publicKeyHex = keyPair.getPublic('hex'); // Uncompressed: 04 + x + y = 130 hex chars
  return { privateKey, publicKeyHex, keyPair };
}

/**
 * Generate 32-byte client nonce (64 hex characters).
 * Required by all TEE providers.
 */
function generateNonce() {
  const bytes = new Uint8Array(32);
  crypto.getRandomValues(bytes);
  return bytesToHex(bytes);
}

/**
 * Fetch and verify TEE attestation for a model.
 */
async function fetchAttestation(modelId) {
  const clientNonce = generateNonce();

  const response = await fetch(
    `${BASE_URL}/tee/attestation?model=${encodeURIComponent(modelId)}&nonce=${clientNonce}`,
    {
      headers: { Authorization: `Bearer ${API_KEY}` },
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Attestation request failed: ${response.status} ${error}`);
  }

  const attestation = await response.json();

  // Verify attestation
  if (attestation.verified !== true) {
    throw new Error(`TEE attestation verification failed: ${JSON.stringify(attestation)}`);
  }

  if (attestation.nonce !== clientNonce) {
    throw new Error('Attestation nonce mismatch - possible replay attack');
  }

  // Get model's public key
  const modelPublicKey = attestation.signing_key || attestation.signing_public_key;
  if (!modelPublicKey) {
    throw new Error('No signing key in attestation response');
  }

  return {
    modelPublicKey,
    signingAddress: attestation.signing_address,
    teeProvider: attestation.tee_provider,
    attestation,
  };
}

/**
 * Normalize public key to uncompressed format with 04 prefix.
 */
function normalizePublicKey(hex) {
  const clean = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (clean.length === 128) return '04' + clean;
  return clean;
}

/**
 * Encrypt a message using ECDH + HKDF + AES-256-GCM.
 */
function encryptMessage(plaintext, modelPublicKeyHex) {
  const ec = new EC('secp256k1');

  // Normalize and parse model's public key
  const normalizedKey = normalizePublicKey(modelPublicKeyHex);
  const modelPublicKey = ec.keyFromPublic(normalizedKey, 'hex');

  // Generate ephemeral key pair for this message
  const ephemeralKeyPair = ec.genKeyPair();

  // ECDH: compute shared secret
  const sharedSecret = ephemeralKeyPair.derive(modelPublicKey.getPublic());
  const sharedSecretBytes = new Uint8Array(sharedSecret.toArray('be', 32));

  // Derive AES key using HKDF-SHA256
  const aesKey = hkdf(sha256, sharedSecretBytes, undefined, HKDF_INFO, 32);

  // Generate random 12-byte nonce
  const nonce = randomBytes(12);

  // Encrypt with AES-256-GCM
  const cipher = gcm(aesKey, nonce);
  const encrypted = cipher.encrypt(new TextEncoder().encode(plaintext));

  // Get ephemeral public key (uncompressed: 65 bytes)
  const ephemeralPublic = new Uint8Array(ephemeralKeyPair.getPublic(false, 'array'));

  // Format: ephemeral_public (65 bytes) + nonce (12 bytes) + ciphertext
  const result = new Uint8Array(65 + 12 + encrypted.length);
  result.set(ephemeralPublic, 0);
  result.set(nonce, 65);
  result.set(encrypted, 77);

  return Buffer.from(result).toString('hex');
}

/**
 * Decrypt an E2EE response chunk.
 */
function decryptChunk(ciphertextHex, clientKeyPair) {
  const raw = hexToBytes(ciphertextHex);

  // Parse components (65 + 12 + ciphertext)
  const serverEphemeralPubKey = raw.slice(0, 65);
  const nonce = raw.slice(65, 77);
  const ciphertext = raw.slice(77);

  const ec = new EC('secp256k1');
  const serverKey = ec.keyFromPublic(Buffer.from(serverEphemeralPubKey));

  // ECDH: compute shared secret
  const sharedSecret = clientKeyPair.derive(serverKey.getPublic());
  const sharedSecretBytes = new Uint8Array(sharedSecret.toArray('be', 32));

  // Derive AES key
  const aesKey = hkdf(sha256, sharedSecretBytes, undefined, HKDF_INFO, 32);

  // Decrypt
  const cipher = gcm(aesKey, nonce);
  const plaintext = cipher.decrypt(ciphertext);

  return new TextDecoder().decode(plaintext);
}

/**
 * Check if string looks like hex-encrypted content.
 */
function isHexEncrypted(s) {
  // Minimum: ephemeral_pub (65) + nonce (12) + tag (16) = 93 bytes = 186 hex
  if (s.length < 186) return false;
  return /^[0-9a-fA-F]+$/.test(s);
}

/**
 * Send an E2EE-encrypted chat message and decrypt the response.
 */
async function chatE2EE(prompt) {
  console.log('🔑 Generating ephemeral key pair...');
  const { publicKeyHex, keyPair } = generateKeyPair();

  console.log('🔍 Fetching TEE attestation...');
  const { modelPublicKey, teeProvider } = await fetchAttestation(MODEL);
  console.log(`✅ TEE attestation verified (provider: ${teeProvider || 'unknown'})`);

  console.log('🔐 Encrypting message...');
  const encryptedContent = encryptMessage(prompt, modelPublicKey);

  const messages = [{ role: 'user', content: encryptedContent }];

  console.log('📤 Sending encrypted request...');
  const response = await fetch(`${BASE_URL}/chat/completions`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      'Content-Type': 'application/json',
      'X-Venice-TEE-Client-Pub-Key': publicKeyHex,
      'X-Venice-TEE-Model-Pub-Key': modelPublicKey,
      'X-Venice-TEE-Signing-Algo': 'ecdsa',
    },
    body: JSON.stringify({
      model: MODEL,
      messages,
      stream: true,
      max_tokens: 500,
    }),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Chat request failed: ${response.status} ${error}`);
  }

  console.log('📥 Decrypting response...\n');
  console.log('-'.repeat(50));

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  let fullContent = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop() || '';

    for (const line of lines) {
      if (!line.startsWith('data: ')) continue;
      const data = line.slice(6).trim();
      if (data === '[DONE]') continue;

      try {
        const chunk = JSON.parse(data);
        const content = chunk.choices?.[0]?.delta?.content;

        if (!content) continue;

        // Decrypt if encrypted
        if (isHexEncrypted(content)) {
          const decrypted = decryptChunk(content, keyPair);
          fullContent += decrypted;
          process.stdout.write(decrypted);
        } else {
          fullContent += content;
          process.stdout.write(content);
        }
      } catch (e) {
        if (!(e instanceof SyntaxError)) {
          console.error(`\n[Decryption error: ${e.message}]`);
        }
      }
    }
  }

  console.log('\n' + '-'.repeat(50));
  console.log('\n🔐 Response decrypted end-to-end');

  return fullContent;
}

/**
 * Prompt user for input.
 */
function askQuestion(query) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) =>
    rl.question(query, (answer) => {
      rl.close();
      resolve(answer);
    })
  );
}

/**
 * Main entry point.
 */
async function main() {
  console.log('='.repeat(60));
  console.log('Venice AI E2EE Chat Example');
  console.log('='.repeat(60));
  console.log(`Model: ${MODEL}`);
  console.log(`API: ${BASE_URL}`);
  console.log();

  let prompt = await askQuestion('Enter your message (or press Enter for default): ');
  prompt = prompt.trim();
  if (!prompt) {
    prompt = 'What is end-to-end encryption and why is it important for privacy?';
  }

  console.log();

  try {
    await chatE2EE(prompt);
  } catch (error) {
    console.error(`\n❌ Error: ${error.message}`);
    process.exit(1);
  }
}

main();

File: examples/e2ee_chat.py
#!/usr/bin/env python3
"""
Venice AI E2EE Chat Example

Complete example of end-to-end encrypted communication with Venice AI API.
Your messages are encrypted client-side before transmission, and responses
are encrypted by the TEE before being sent back—Venice never sees plaintext.

Requirements:
    pip install cryptography ecdsa requests

Usage:
    export VENICE_API_KEY="your-api-key"
    python e2ee_chat.py
"""

import os
import json
import secrets
import requests
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from ecdsa import SECP256k1, VerifyingKey, SigningKey

# Configuration
API_KEY = os.environ.get('VENICE_API_KEY')
BASE_URL = os.environ.get('VENICE_API_BASE_URL', 'https://api.venice.ai/api/v1')
MODEL = 'e2ee-qwen3-5-122b-a10b'  # E2EE-capable model
HKDF_INFO = b'ecdsa_encryption'

if not API_KEY:
    print('Error: VENICE_API_KEY environment variable not set')
    print('Usage: export VENICE_API_KEY="your-api-key"')
    exit(1)


def generate_key_pair():
    """Generate ephemeral secp256k1 key pair for E2EE session."""
    private_key = SigningKey.generate(curve=SECP256k1)
    public_key = private_key.get_verifying_key()
    # Uncompressed public key format: 04 || x (32 bytes) || y (32 bytes) = 65 bytes
    public_key_bytes = b'\x04' + public_key.to_string()
    return private_key, public_key_bytes.hex()


def fetch_attestation(model_id: str) -> dict:
    """Fetch and verify TEE attestation for a model."""
    # Generate 32-byte nonce (64 hex chars) - required by all TEE providers
    client_nonce = secrets.token_hex(32)
    
    response = requests.get(
        f'{BASE_URL}/tee/attestation',
        params={'model': model_id, 'nonce': client_nonce},
        headers={'Authorization': f'Bearer {API_KEY}'},
        timeout=30
    )
    response.raise_for_status()
    attestation = response.json()
    
    # Verify attestation
    if attestation.get('verified') != True:
        raise ValueError(f"TEE attestation verification failed: {attestation}")
    
    if attestation.get('nonce') != client_nonce:
        raise ValueError('Attestation nonce mismatch - possible replay attack')
    
    # Get model's public key for encryption
    model_public_key = attestation.get('signing_key') or attestation.get('signing_public_key')
    if not model_public_key:
        raise ValueError('No signing key in attestation response')
    
    return {
        'model_public_key': model_public_key,
        'signing_address': attestation.get('signing_address'),
        'tee_provider': attestation.get('tee_provider'),
        'attestation': attestation
    }


def encrypt_message(plaintext: str, model_public_key_hex: str) -> str:
    """Encrypt a message using ECDH + HKDF + AES-256-GCM."""
    # Normalize public key (add 04 prefix if missing)
    key_hex = model_public_key_hex
    if not key_hex.startswith('04') and len(key_hex) == 128:
        key_hex = '04' + key_hex
    
    model_public_key_bytes = bytes.fromhex(key_hex)
    
    # Parse model's public key (skip 04 prefix)
    model_verifying_key = VerifyingKey.from_string(
        model_public_key_bytes[1:], 
        curve=SECP256k1
    )
    
    # Generate ephemeral key pair for this message
    ephemeral_private = SigningKey.generate(curve=SECP256k1)
    ephemeral_public = ephemeral_private.get_verifying_key()
    
    # ECDH: compute shared secret
    shared_point = model_verifying_key.pubkey.point * ephemeral_private.privkey.secret_multiplier
    shared_secret = shared_point.x().to_bytes(32, 'big')
    
    # Derive AES key using HKDF-SHA256
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=None,
        info=HKDF_INFO,
    )
    aes_key = hkdf.derive(shared_secret)
    
    # Generate random 12-byte nonce
    nonce = os.urandom(12)
    
    # Encrypt with AES-256-GCM
    aesgcm = AESGCM(aes_key)
    ciphertext = aesgcm.encrypt(nonce, plaintext.encode('utf-8'), None)
    
    # Format: ephemeral_public (65 bytes) + nonce (12 bytes) + ciphertext
    ephemeral_public_bytes = b'\x04' + ephemeral_public.to_string()
    result = ephemeral_public_bytes + nonce + ciphertext
    
    return result.hex()


def decrypt_chunk(ciphertext_hex: str, client_private_key: SigningKey) -> str:
    """Decrypt an E2EE response chunk."""
    raw = bytes.fromhex(ciphertext_hex)
    
    # Parse components (65 + 12 + ciphertext)
    server_ephemeral_pub = raw[:65]
    nonce = raw[65:77]
    ciphertext = raw[77:]
    
    # Parse server's ephemeral public key (skip 04 prefix)
    server_verifying_key = VerifyingKey.from_string(
        server_ephemeral_pub[1:],
        curve=SECP256k1
    )
    
    # ECDH: compute shared secret
    shared_point = server_verifying_key.pubkey.point * client_private_key.privkey.secret_multiplier
    shared_secret = shared_point.x().to_bytes(32, 'big')
    
    # Derive AES key
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=None,
        info=HKDF_INFO,
    )
    aes_key = hkdf.derive(shared_secret)
    
    # Decrypt
    aesgcm = AESGCM(aes_key)
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    
    return plaintext.decode('utf-8')


def is_hex_encrypted(s: str) -> bool:
    """Check if string looks like hex-encrypted content."""
    # Minimum: ephemeral_pub (65) + nonce (12) + tag (16) = 93 bytes = 186 hex
    if len(s) < 186:
        return False
    return all(c in '0123456789abcdefABCDEF' for c in s)


def chat_e2ee(prompt: str) -> str:
    """Send an E2EE-encrypted chat message and decrypt the response."""
    print('🔑 Generating ephemeral key pair...')
    client_private_key, client_public_key_hex = generate_key_pair()
    
    print('🔍 Fetching TEE attestation...')
    attestation_result = fetch_attestation(MODEL)
    model_public_key = attestation_result['model_public_key']
    print(f'✅ TEE attestation verified (provider: {attestation_result.get("tee_provider", "unknown")})')
    
    print('🔐 Encrypting message...')
    encrypted_content = encrypt_message(prompt, model_public_key)
    
    messages = [{'role': 'user', 'content': encrypted_content}]
    
    print('📤 Sending encrypted request...')
    response = requests.post(
        f'{BASE_URL}/chat/completions',
        headers={
            'Authorization': f'Bearer {API_KEY}',
            'Content-Type': 'application/json',
            'X-Venice-TEE-Client-Pub-Key': client_public_key_hex,
            'X-Venice-TEE-Model-Pub-Key': model_public_key,
            'X-Venice-TEE-Signing-Algo': 'ecdsa'
        },
        json={
            'model': MODEL,
            'messages': messages,
            'stream': True,
            'max_tokens': 500,
        },
        stream=True,
        timeout=60
    )
    response.raise_for_status()
    
    print('📥 Decrypting response...\n')
    print('-' * 50)
    
    full_content = ''
    for line in response.iter_lines():
        if not line:
            continue
        
        line_str = line.decode('utf-8')
        if not line_str.startswith('data: ') or '[DONE]' in line_str:
            continue
        
        try:
            chunk = json.loads(line_str[6:])
            content = chunk.get('choices', [{}])[0].get('delta', {}).get('content', '')
            
            if not content:
                continue
            
            # Decrypt if encrypted
            if is_hex_encrypted(content):
                decrypted = decrypt_chunk(content, client_private_key)
                full_content += decrypted
                print(decrypted, end='', flush=True)
            else:
                full_content += content
                print(content, end='', flush=True)
        except json.JSONDecodeError:
            pass
        except Exception as e:
            print(f'\n[Decryption error: {e}]')
    
    print('\n' + '-' * 50)
    print('\n🔐 Response decrypted end-to-end')
    
    return full_content


def main():
    print('=' * 60)
    print('Venice AI E2EE Chat Example')
    print('=' * 60)
    print(f'Model: {MODEL}')
    print(f'API: {BASE_URL}')
    print()
    
    prompt = input('Enter your message (or press Enter for default): ').strip()
    if not prompt:
        prompt = 'What is end-to-end encryption and why is it important for privacy?'
    
    print()
    chat_e2ee(prompt)


if __name__ == '__main__':
    main()
